### PR TITLE
[BD-46] feat: eslint config upgrade, refactor

### PR DESCRIPTION
**Pull request by monteri**
## Description

Most of hooks are updated due to ESLint rules and some are `eslint-disabled` because of big amount of refactor required. `ProductTour` component was refactored because of inefficient `useEffect` usage.

JIRA: https://openedx.atlassian.net/browse/PAR-804

### Deploy Preview

[Preview](https://deploy-preview-1263--paragon-openedx.netlify.app/)

## Merge Checklist

* [x] If your update includes visual changes, have they been reviewed by a designer? Send them a link to the Netlify deploy preview, if applicable.
* [x] Does your change adhere to the documented [style conventions](https://github.com/openedx/paragon/blob/master/docs/decisions/0012-css-styling-conventions)?
* [ ] Do any prop types have missing descriptions in the Props API tables in the documentation site (check deploy preview)?
* [x] Were your changes tested using all available themes (see theme switcher in the header of the deploy preview, under the "Settings" icon)?
* [x] Is there adequate test coverage for your changes?
* [ ] Consider whether this change needs to reviewed/QA'ed for accessibility (a11y). If so, please add `wittjeff` and `adamstankiewicz` as reviewers on this PR.

## Post-merge Checklist

* [ ] Verify your changes were released to [NPM](https://www.npmjs.com/package/@edx/paragon) at the expected version.
* [ ] If you'd like, [share](https://github.com/openedx/paragon/discussions/new?category=show-and-tell) your contribution in [#show-and-tell](https://github.com/openedx/paragon/discussions/categories/show-and-tell).
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @monteri! I've created [BLENDED-1222](https://openedx.atlassian.net/browse/BLENDED-1222) to keep track of it in Jira.   More details are on the [BD-46](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3134685276/BD-46+Paragon+Blended+Support+and+Maintenance) project page.


When this pull request is ready, tag your edX technical lead.

File: src/Sticky/index.jsx [hunk 1]
```diff
@@ -41,10 +41,11 @@ const Sticky = React.forwardRef(({
       observer.observe(resolvedRef.current);
 
       return () => {
+        // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (viktorrusakov) [inline src/Sticky/index.jsx:44]: Eslint suggests that we should store `resolvedRef.current` as a variable somewhere inside the hook because ref's value might change before this function will actually be called, that's why it is better to remember the value. Maybe we could do it instead of disabling the rule here?

File: src/ProductTour/index.jsx [hunk 2]
```diff
@@ -5,49 +5,51 @@ import Checkpoint from './Checkpoint';
...
-  const [currentCheckpointData, setCurrentCheckpointData] = useState(null);
+  const {
+    enabled, checkpoints = [], startingIndex, onEscape, onEnd, onDismiss: tourOnDismiss,
+    advanceButtonText: tourAdvanceButtonText, dismissButtonText: tourDismissButtonText,
+    endButtonText: tourEndButtonText,
+  } = tourValue || {};
+  const [currentCheckpointData, setCurrentCheckpointData] = useState([]);
   const [index, setIndex] = useState(0);
-  const [isTourEnabled, setIsTourEnabled] = useState(!!tourValue);
+  const [isTourEnabled, setIsTourEnabled] = useState(false);
```
Comment (viktorrusakov) [inline src/ProductTour/index.jsx:15]: [curious] why did you change initial value to be always `false`?

File: src/Form/FormGroupContext.jsx [hunk 3]
```diff
@@ -63,6 +63,7 @@ const FormGroupContextProvider = ({
   ]);
 
   const getLabelProps = (labelProps) => {
+    // eslint-disable-next-line react-hooks/rules-of-hooks
     const id = useRegisteredLabellerId(labelProps?.id);
```
Comment (adamstankiewicz) [inline src/Form/FormGroupContext.jsx:67]: It seems like `useRegisteredLabellerId` itself is not really a hook on its own, but rather a function returned by a custom hook. Is this getting flagged as an eslint error just due to the variable prefix `use`? Would renaming the function in the return value of the `useIdList` above solve the issue?

Same question for the disabled line below, too.

File: src/DataTable/index.jsx [hunk 4]
```diff
@@ -111,10 +111,12 @@ function DataTable({
   // any state changes to current row selections as we don't want to re-fetch data whenever row(s) are selected.
   const tableStateWithoutSelections = { ...instance.state };
   delete tableStateWithoutSelections.selectedRowIds;
+
   useEffect(() => {
     if (fetchData) {
       fetchData(tableStateWithoutSelections);
     }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
[the remark this answers is not in the document]
Comment (monteri) [reply src/DataTable/index.jsx:119]: The point is that in the comment above the `tableStateWithoutSelections`:

> Call ``fetchData`` whenever the state of the table changes (e.g., page change, sort or filter applied) but ignore

This means we want this `useEffect` to work every time the `instance.state` changes. If we just pass `tableStateWithoutSelections` without `JSON.stringify`, any variables changes inside won't be detected due to not deep comparison of `Objects`. The same thing will be for `useMemo` because we would need `JSON.stringify` for `instance.state`. Otherwise changes to state won't trigger changes in the `tableStateWithoutSelections` variable.

File: src/DataTable/selection/ControlledSelect.jsx [hunk 5]
```diff
@@ -24,10 +26,10 @@ const ControlledSelect = ({ row }) => {
         dispatch(addSelectedRowAction(row, itemCount));
       }
     },
-    [itemCount, row],
+    [itemCount, row, dispatch],
   );
 
-  const toggleRowSelectedProps = useMemo(() => row.getToggleRowSelectedProps(), [row.getToggleRowSelectedProps]);
+  const toggleRowSelectedProps = useMemo(() => getToggleRowSelectedProps(), [getToggleRowSelectedProps]);
```
[the remark this answers is not in the document]
Comment (monteri) [reply src/DataTable/selection/ControlledSelect.jsx:32]: Yes, I checked, `useMemo` doesn't change anything so I removed it.

File: src/DataTable/utils/getVisibleColumns.jsx [hunk 6]
```diff
@@ -32,7 +32,8 @@ export const selectColumn = {
   // Proptypes disabled as this prop is passed in separately
   /* eslint-disable react/prop-types */
   Cell: ({ row }) => {
-    const toggleRowSelectedProps = useMemo(() => row.getToggleRowSelectedProps(), [row.getToggleRowSelectedProps]);
+    const { getToggleRowSelectedProps } = row || {};
+    const toggleRowSelectedProps = useMemo(() => getToggleRowSelectedProps(), [getToggleRowSelectedProps]);
```
[the remark this answers is not in the document]
Comment (monteri) [reply src/DataTable/utils/getVisibleColumns.jsx:36]: Yes, I checked, useMemo doesn't change anything so I removed it.

File: src/Form/FormGroupContext.jsx (hunk 3 above)
Comment (monteri) [reply src/Form/FormGroupContext.jsx]: Yes renaming solved the issue, thank you

**Review by adamstankiewicz — APPROVED** [review]
LGTM! 🚀

**Comment by openedx-webhooks** [comment]
@monteri 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/paragon&entry.867055334=https://github.com/openedx/paragon/pull/1263&entry.1484655318=https://github.com/monteri&entry.752974735=2022-05-03+12:42&entry.1917517419=2022-05-20+10:35&entry.2133058324=Yes) so we can improve your experience in the future.
